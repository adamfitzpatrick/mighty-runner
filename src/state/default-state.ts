import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
