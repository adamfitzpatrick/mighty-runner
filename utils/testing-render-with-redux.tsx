import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
