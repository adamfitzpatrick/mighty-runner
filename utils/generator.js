const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
