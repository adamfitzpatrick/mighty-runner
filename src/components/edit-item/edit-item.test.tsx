import * as React from 'react'
import { cleanup, render, fireEvent } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import EditItem from '.'
import { Stat } from '@models'
import { AppState } from '@state/default-state'

describe('EditItem component', () => {
  let appState: AppState
  let item: Stat
  let changeSpy: jest.Mock<Stat>
  let doneSpy: jest.Mock<undefined>
  let renderFunc: (item: Stat, changeHandler: (item: Stat) => void) => JSX.Element
  let sut: RenderWithRedux

  beforeEach(() => {
    appState = {
      effects: [{
        id: '1',
        name: 'n',
        description: 'd',
        target: [ 'a', 'b' ],
        active: true,
        value: 1
      }]
    } as AppState
    item = {
      id: '1',
      name: 'n',
      shortName: 's',
      description: 'd',
      asEffectTarget: [ 'a', 'b' ],
      effects: [ '1' ],
      value: { thing: 1 }
    }
    changeSpy = jest.fn()
    doneSpy = jest.fn()
    renderFunc = (item: Stat, changeHandler: (item: Stat) => void) => {
      const onChange = () => changeHandler(item)
      return <input
        data-testid='rendered-in-edit-item.input'
        value='value'
        onChange={ onChange }
      />
    }
    sut = renderWithRedux(<EditItem
      item={ item }
      render={ renderFunc }
      changeHandler={ changeSpy }
      done={ doneSpy }
    />, appState)
  })

  test('should render correctly', () => {
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  test('should pass on the change handler from the parent component', () => {
    const innerInput = sut.getByTestId('rendered-in-edit-item.input')
    fireEvent.change(innerInput, { target: { value: 'change' } })
    expect(changeSpy).toHaveBeenCalled()
  })

  test('should call done when the done button is clicked', () => {
    const button = sut.getByTestId('Done.button.component')
    fireEvent.click(button)
    expect(doneSpy).toHaveBeenCalled()
  })

  test('should render correctly when item has no related effects', () => {
    cleanup()
    delete item.effects
    let noEffects = render(<EditItem
      item={ item }
      render={ renderFunc }
      changeHandler={ changeSpy }
      done={ doneSpy }
    />)
    expect(noEffects.container.innerHTML).toMatchSnapshot()
  })
})
