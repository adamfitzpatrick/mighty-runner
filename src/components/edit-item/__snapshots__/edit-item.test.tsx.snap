// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItem component should render correctly 1`] = `"<div><input data-testid=\\"rendered-in-edit-item.input\\" value=\\"value\\"><div data-testid=\\"editable-effect.component\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"n\\"></label><label><span>Description</span><input data-testid=\\"Description.input.component\\" type=\\"text\\" value=\\"d\\"></label><label>Target<input data-testid=\\"0.array.input\\" type=\\"text\\" value=\\"a\\"></label><label><span class=\\"additionalLabel\\">Target</span><input data-testid=\\"1.array.input\\" type=\\"text\\" value=\\"b\\"></label><label><span>Value</span><input data-testid=\\"Value.input.component\\" type=\\"text\\" value=\\"1\\"></label><label><span>Active</span><input data-testid=\\"Active.input.component\\" type=\\"checkbox\\" value=\\"\\" checked=\\"\\"></label></div><button data-testid=\\"Done.button.component\\">Done</button></div>"`;

exports[`EditItem component should render correctly when item has no related effects 1`] = `"<div><input data-testid=\\"rendered-in-edit-item.input\\" value=\\"value\\"><button data-testid=\\"Done.button.component\\">Done</button></div>"`;
